File: day7/typescript/part_2.ts
import * as fs from 'fs';

const input = fs.readFileSync('input.txt', 'utf-8').trim().split('\n');
let total = 0;

for (const line of input) {
    const [testValue, nums] = line.split(':');
    const numbers = nums.trim().split(/\s+/).map(Number);
    const opsCount = numbers.length - 1;
    let found = false;

    for (let i = 0; i < (1 << (2 * opsCount)) && !found; i++) {
        let result = numbers[0];
        let valid = true;
        for (let j = 0; j < opsCount && valid; j++) {
            const op = (i >> (2 * j)) & 3;
            switch (op) {
                case 0:
                    result += numbers[j + 1];
                    break;
                case 1:
                    result -= numbers[j + 1];
                    break;
                case 2:
                    result *= numbers[j + 1];
                    break;
                case 3:
                    if (numbers[j + 1] === 0 || result % numbers[j + 1] !== 0) {
                        valid = false;
                    } else {
                        result /= numbers[j + 1];
                    }
                    break;
            }
        }
        if (valid && result === Number(testValue)) {
            total += Number(testValue);
            found = true;
        }
    }
}

console.log(total);

File: day7/ruby/part_2.rb
total = 0

File.foreach('input.txt') do |line|
  test_value, nums = line.split(':')
  test_value = test_value.to_i
  numbers = nums.split.map(&:to_i)
  
  ops_count = numbers.size - 1
  found = false
  (0...4**ops_count).each do |i|
    break if found
    result = numbers[0]
    valid = true
    ops_count.times do |j|
      op = (i >> (2 * j)) & 3
      case op
      when 0
        result += numbers[j + 1]
      when 1
        result -= numbers[j + 1]
      when 2
        result *= numbers[j + 1]
      when 3
        if numbers[j + 1] == 0 || result % numbers[j + 1] != 0
          valid = false
          break
        end
        result /= numbers[j + 1]
      end
    end
    if valid && result == test_value
      total += test_value
      found = true
    end
  end
end

puts total
